core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -65,17 +68,19 @@ class TransactionCoordinator(val brokerId: Int,
       val pid: Long = pidManager.getNewPid()
 
       responseCallback(InitPidResult(pid, epoch = 0, Errors.NONE))
-    } else if(!logManager.isCoordinatorFor(transactionalId)) {
+    } else if (!txnManager.isCoordinatorFor(transactionalId)) {
       // check if it is the assigned coordinator for the transactional id
-      responseCallback(initPidError(Errors.NOT_COORDINATOR))
+      responseCallback(initTransactionError(Errors.NOT_COORDINATOR))
+    } else if (!txnManager.isCoordinatorLoadingInProgress(transactionalId)) {
+      responseCallback(initTransactionError(Errors.COORDINATOR_LOAD_IN_PROGRESS))
     } else {
       // only try to get a new pid and update the cache if the transactional id is unknown
-      getPidMetadata(transactionalId) match {
+      txnManager.getTransaction(transactionalId) match {
         case None =>
           val pid: Long = pidManager.getNewPid()
           // TODO: check transactionTimeoutMs is not larger than the broker configured maximum allowed value
-          val newMetadata: PidMetadata = new PidMetadata(pid, epoch = 0, transactionTimeoutMs)
-          val metadata = addPidMetadata(transactionalId, newMetadata)
+          val newMetadata: TransactionMetadata = new TransactionMetadata(pid, epoch = 0, transactionTimeoutMs)
+          val metadata = txnManager.addTransaction(transactionalId, newMetadata)
 
           // there might be a concurrent thread that has just updated the mapping
```

Comment:
I think it is following what we have described in the design doc: whenever a initPid request is sent with a known txn id, no matter if they are from the same producer instance or not, and no matter if it is very close to the previous one, it will always cause the previous returned PID/epoch to be fenced.